File: FastOS/kernel/src/drivers/timer.rs
// ============================================================
// FastOS — PIT Timer Driver
// ============================================================
// Programmable Interval Timer (8253/8254).
// Channel 0 connected to IRQ 0 for system tick.
// ============================================================

use crate::arch::x86_64::port;

/// PIT I/O ports
const PIT_CHANNEL0: u16 = 0x40;
const PIT_COMMAND: u16  = 0x43;

/// PIT base frequency: 1,193,182 Hz
const PIT_BASE_FREQ: u32 = 1_193_182;

/// System tick counter
static mut TICKS: u64 = 0;
static mut FREQUENCY: u32 = 0;

/// Initialize PIT at the given frequency (Hz)
pub fn init(freq: u32) {
    let divisor = PIT_BASE_FREQ / freq;
    let divisor = if divisor > 65535 { 65535 } else if divisor < 1 { 1 } else { divisor };

    unsafe { FREQUENCY = freq; }

    // Command: channel 0, lobyte/hibyte, rate generator (mode 2)
    port::outb(PIT_COMMAND, 0x36);

    // Send divisor (low byte first, then high byte)
    port::outb(PIT_CHANNEL0, (divisor & 0xFF) as u8);
    port::outb(PIT_CHANNEL0, ((divisor >> 8) & 0xFF) as u8);
}

/// Called from timer IRQ handler — increment tick counter
pub fn tick() {
    unsafe { TICKS += 1; }
}

/// Get current tick count
pub fn get_ticks() -> u64 {
    unsafe { TICKS }
}

/// Get configured frequency
pub fn get_frequency() -> u32 {
    unsafe { FREQUENCY }
}

/// Get uptime in milliseconds
pub fn uptime_ms() -> u64 {
    let freq = get_frequency() as u64;
    if freq == 0 { return 0; }
    (get_ticks() * 1000) / freq
}

/// Get uptime in seconds
pub fn uptime_secs() -> u64 {
    let freq = get_frequency() as u64;
    if freq == 0 { return 0; }
    get_ticks() / freq
}

/// Busy-wait for approximately `ms` milliseconds
pub fn sleep_ms(ms: u64) {
    let target = get_ticks() + (ms * get_frequency() as u64) / 1000;
    while get_ticks() < target {
        crate::arch::x86_64::cpu::hlt();
    }
}

/// Busy-wait for approximately `us` microseconds (rough)
pub fn sleep_us(us: u64) {
    // At 1000 Hz, minimum resolution is 1ms
    // For sub-ms delays, use I/O port delay loop
    let loops = us / 10;
    for _ in 0..loops {
        port::io_wait();
    }
}

File: FastOS/kernel/drivers/timer.c
/* FastOS v2.1 — PIT Timer Driver (Intel 8253/8254)
 * ADead-BIB Native OS
 *
 * PIT (Programmable Interval Timer):
 *   Channel 0: System timer (IRQ0 → vector 0x20 after PIC remap)
 *   Channel 1: DRAM refresh (legacy, unused)
 *   Channel 2: PC speaker
 *
 * Ports:
 *   0x40: Channel 0 data
 *   0x41: Channel 1 data
 *   0x42: Channel 2 data
 *   0x43: Mode/Command register
 *
 * Command byte (port 0x43):
 *   Bits 7-6: Channel (00=ch0)
 *   Bits 5-4: Access mode (11=lo/hi byte)
 *   Bits 3-1: Mode (011=square wave, mode 3)
 *   Bit 0:    BCD (0=binary)
 *   → 0x36 = channel 0, lo/hi, mode 3, binary
 *
 * Base frequency: 1,193,182 Hz
 * Divisor for ~100 Hz: 11932 (0x2E9C)
 * Divisor for ~1000 Hz: 1193 (0x04A9)
 *
 * Inline init for kernel_main():
 *   __outb(0x43, 0x36);           // Channel 0, lo/hi, mode 3
 *   __outb(0x40, 0x9C);           // Divisor low byte (11932)
 *   __outb(0x40, 0x2E);           // Divisor high byte
 *
 * Tick counter (global variable):
 *   int timer_ticks = 0;
 *   // In IRQ0 handler: timer_ticks = timer_ticks + 1;
 *   // Uptime seconds: timer_ticks / 100
 *
 * Without IDT/IRQ: use TSC (RDTSC) for timing
 *   __rdtsc() returns 64-bit timestamp counter
 *   Ryzen 5 5600X base: 3.7 GHz → 3,700,000,000 ticks/sec
 */
